watch/source.go

```diff
@@ -53,20 +58,33 @@ func NewSource(poll SourcePollFn, logger xlog.Logger) Source {
 type source struct {
 	sync.RWMutex
 
-	poll   SourcePollFn
-	w      Watchable
-	logger xlog.Logger
-	closed bool
+	poll        SourcePollFn
+	w           Watchable
+	closed      bool
+	initialized bool
+	logger      xlog.Logger
 }
 
 func (s *source) run() {
 	for !s.isClosed() {
 		data, err := s.poll()
+		if err == ErrSourceClosed {
+			s.logger.Errorf("Upstream source is closed")
+			s.Close()
+			return
+		}
 		if err != nil {
```

Comment:
The backoff/jitter could be left in the impl of the pollFn, but probably safer to guard it here too, I'm ok either way.